geometry.cpp

```diff
@@ -160,11 +160,76 @@ drawvec remove_noop(drawvec geom, int type, int shift) {
 	return out;
 }
 
-double get_area(drawvec &geom, size_t i, size_t j) {
+double get_area_scaled(const drawvec &geom, size_t i, size_t j) {
+	const double max_exact_double = (double) ((1LL << 53) - 1);
+
+	// keep scaling the geometry down until we can calculate its area without overflow
+	for (long long scale = 2; scale < (1LL << 30); scale *= 2) {
+		long long bx = geom[i].x;
+		long long by = geom[i].y;
+		bool again = false;
+
+		// https://en.wikipedia.org/wiki/Shoelace_formula
+		double area = 0;
+		for (size_t k = i; k < j; k++) {
+			area += (double) ((geom[k].x - bx) / scale) * (double) ((geom[i + ((k - i + 1) % (j - i))].y - by) / scale);
+			if (std::fabs(area) >= max_exact_double) {
+				again = true;
+				break;
+			}
+			area -= (double) ((geom[k].y - by) / scale) * (double) ((geom[i + ((k - i + 1) % (j - i))].x - bx) / scale);
+			if (std::fabs(area) >= max_exact_double) {
+				again = true;
+				break;
+			}
+		}
+
+		if (again) {
+			continue;
+		} else {
+			area /= 2;
+			return area * scale * scale;
+		}
+	}
+
+	fprintf(stderr, "get_area_scaled: can't happen\n");
+	exit(EXIT_IMPOSSIBLE);
+}
+
+double get_area(const drawvec &geom, size_t i, size_t j) {
+	const double max_exact_double = (double) ((1LL << 53) - 1);
+
+	// Coordinates in `geom` are 40-bit integers, so there is no good way
+	// to multiply them without possible precision loss. Since they probably
+	// do not use the full precision, shift them nearer to the origin so
+	// their product is more likely to be exactly representable as a double.
+	//
+	// (In practice they are actually 34-bit integers: 32 bits for the
+	// Mercator world plane, plus another two bits so features can stick
+	// off either the left or right side. But that is still too many bits
+	// for the product to fit either in a 64-bit long long or in a
+	// double where the largest exact integer is 2^53.)
+	//
+	// If the intermediate calculation still exceeds 2^53, start trying to
+	// recalculate the area by scaling down the geometry. This will not
+	// produce as accurate an area, but it will still be close, and the
+	// sign will be correct, which is more important, since the sign
+	// determines the winding order of the rings.
+
+	long long bx = geom[i].x;
+	long long by = geom[i].y;
+
+	// https://en.wikipedia.org/wiki/Shoelace_formula
```

Comment:
If all coordinates are integers, can we use [Pick's Theorem](https://en.wikipedia.org/wiki/Pick%27s_theorem) instead of the Shoelace Formula? It's a couple of additions and a division so we shouldn't get precision problems even with 40 bit values. 
It assumes polygons to be non-intersecting and have no holes but if we need to account for holes we can just compute their area and substract it.

Reply:
Thanks! I didn't know about that one, and will give it a try. This calculation is applied only to one ring at a time, so intersections shouldn't be a problem.

Reply:
Maybe I'm misunderstanding, but is there a way to do this without effectively rasterizing the polygon to get the count of interior integer coordinate pairs?

Reply:
I think you are right, I was remembering a way to do that without rasterizing, but looking into it, it involves triangulating the polygon and using the triangle areas to compute them, at which point it's not really useful because you'd already have the area you want.
I assume rasterizing the polygon is not an option, is it?